Adding option to disable removal of "{{user}}:" and "{{char}}:" in generateRaw()

qvink opened the pull request:
### Description:
In `cleanUpMessage()`, the lines for removing "{{user}}:" or "{{char}}:" at the beginning of responses would instead remove the entire response. This was because they evaluated to `message.substring(0, 0)`. I have modified it so that it only removes the name at the beginning, i.e. `message.substring(index+length)`

### Motivation:
This fixes a bug where `generateRaw()` would mistakenly return an empty string when "{{user}}:" or "{{char}}:" appeared at the beginning and the corresponding user settings were disabled ("Show {{char}}: in responses" and "Show {{user}}: in responses").

I am not *fully* aware of what secondary effects this might have, but this change does fix that bug.

## Checklist:
- [x] I have read the [Contributing guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

qvink committed:
Fix for removal of "{{user}}:" from generations

Wolfsblvt commented on public/script.js:
```diff
@@ -5971,10 +5971,10 @@ export function cleanUpMessage(getMessage, isImpersonate, isContinue, displayInc
 
     if (nameToTrim && getMessage.indexOf(`${nameToTrim}:`) == 0) {
-        getMessage = getMessage.substring(0, getMessage.indexOf(`${nameToTrim}:`));
+        getMessage = getMessage.substring(getMessage.indexOf(`${nameToTrim}:`)+nameToTrim.length);
```
This is still a bit weird.
I think it needs to check if it **starts** with the nameToTrim, so that it is at index 0. Which makes more sense.
Which then in return means you don't need to get the index again, you can just use 0, like before.

Also, I think it needs to be `nameToTrim.length + 1`, because of the colon. And then there might still be whitespaces at the beginning, so it should trim those as well.

qvink replied:
I am confused - we *don't* want to use 0. Substring returns the text from the first argument to the second, or to the end of the string if no second arg is given. If you do `.substring(0, x)`, then you are returning the text from the start to x, not removing it.

I agree about the extra length and whitespace.

Wolfsblvt replied:
You are calculating `getMessage.indexOf(`${nameToTrim}:`)` + `nameToTrim.length`.
The first one always defaults to 0, according to the strict rules. No need to do that.

qvink committed:
removing redundant .indexOf(). Adding extra length to account for the colon.

Cohee1207 reviewed (changes requested):
You probably have to pass _another argument_ to allow name prefixes not to get trimmed.

Cohee1207 commented:
And if we're gonna add another `cleanUpMessage` arg, might as well switch to using object args :)

See recent changes to `saveReply` and `saveChat`

qvink committed:
cleanUpMessages() now uses object args. Adding option to trim names.

qvink commented:
Alright I've made `cleanUpMessage()` use object arguments and updated all uses to use them. I have also added a new argument to toggle trimming names from the beginning, true by default.

However, I noticed that there is another block of code later in the function that appears to do the exact same thing, just without the newline case. Are both these blocks intended to be there or can we remove the first one?

qvink committed:
formatting

qvink commented on public/script.js:
```diff
@@ -6035,13 +6101,15 @@ export function cleanUpMessage(getMessage, isImpersonate, isContinue, displayInc
-        ? (!power_user.allow_name1_display ? name1 : '')
-        : (!power_user.allow_name2_display ? name2 : '');
+    if (trimNames) {
+        const nameToTrim2 = isImpersonate
```
This is the second block which trims names from the beginning, though it appears to do the exact same thing as the first but without handling the newline case.

Cohee1207 commented:
> I noticed that there is another block of code later in the function that appears to do the exact same thing, just without the newline case.

It's okay.

> or can we remove the first one?

Please don't.

qvink commented:
Hmm actually the second one isn't the same, I think the second one has `name1` and `name2` flipped. In the first block, `name2` is used if `isImpersonate` is true. In the second block, `name1` is used in that case. Is this intentional?

Wolfsblvt commented:
The answer to that is very likely yes.

qvink commented:
Terrifying, thank you

Cohee1207 commented:
Yes, both cases exist to trim leaked replies. What to trim depends on the context.

Cohee1207 committed:
Sorry we don't know why lint didn't work

Cohee1207 commented on public/script.js:
```diff
@@ -5961,21 +6020,28 @@ export function cleanUpMessage(getMessage, isImpersonate, isContinue, displayInc
+        if (nameToTrim && startIndex >= 0) {
+            getMessage = getMessage.substring(startIndex + nameToTrim.length + 2);
+        }
+        getMessage.trimStart();
```
This trimStart will not do anything since it's not assigned to a variable

Cohee1207 commented on public/script.js:
```diff
@@ -5961,21 +6020,28 @@ export function cleanUpMessage(getMessage, isImpersonate, isContinue, displayInc
-    }
+        // get text from after the name (and colon) to the end of the string
+        if (nameToTrim && getMessage.indexOf(`${nameToTrim}:`) === 0) {
+            getMessage = getMessage.substring(nameToTrim.length + 1);
```
I think this block is entirely wrong now. This code was there since the beginning, and I think it's meant to nullify the string if it starts with impersonated reply. Now it just trims the name, but the impersonation stays.

Wolfsblvt replied:
Now that you say it...
But still, if that's the case, it was still done wrong. The first if block works like this, yes. But the second one checks if there somewhere is a new line, starting with the name to trim, and then just removes the whole message - instead of everything starting from that position.

Cohee1207 replied:
1. Message starst with a name to trim - nuke the entire message
2. Name to trim occurs on a different line - nuke the message from that point

Wolfsblvt replied:
Makes sense to me. Code did something different before. So... maybe this whole PR needs a start-over?

Cohee1207 replied:
I like the direction it's taking, the actual code in `if` conditions just needs to be as it used to be.

qvink replied:
If we revert the actual function of `cleanUpMessage()`, then the bug described originally wouldn't be addressed - that the entire response gets deleted if "{{char/user}}:" is present at the beginning.

Wolfsblvt replied:
If `{{user}}:` is present at the beginning the entire response **should** get deleted. This is impersonation. This is not expected, and canceled by stopping strings - and this code - on purpose.

Cohee1207 replied:
To resolve your original issue, `trimNames` should be an optional argument of `generateRaw` too

qvink committed:
Reverting cleanUpMessage() functionality changes.

qvink committed:
Adding missing doc string

qvink commented:
Alright, I think I now understand the intent here. There are now two new options for `cleanUpMessage()`, `trimNames` and `trimWrongNames`. 

`trimNames=true` will remove "{{char}}:" and "{{user}}:" from the beginning if the appropriate settings are not enabled.

`trimWrongNames=true` will remove the entire responses if the *wrong* name is at the beginning, exactly the same as it functioned before, but now explicitly setting `getMessage = ''` instead of using `getMessage.substring(0,0)`. 

In `generateRaw()`, there is one argument to toggle both of these at the same time, though let me know if you'd rather they be separated there too.

It is unclear to me why the functionality for deleting the entire response is tied to the power user settings "Show {{char}}: in responses" and "Show {{user}} in responses", but that is how it functioned originally.

Cohee1207 commented on public/script.js:
```diff
@@ -6020,26 +6023,33 @@ export function cleanUpMessage({ getMessage, isImpersonate, isContinue, displayI
-            getMessage = getMessage.substring(startIndex + nameToTrim.length + 2);
+        if (wrongName) {
+            // If the message starts with the wrong name, delete the entire response
+            let startIndex = getMessage.indexOf(`${wrongName}`);
```
Important distinction: previously it was `${name}:` with a colon.

qvink replied:
fixed

Cohee1207 commented on public/script.js:
```diff
@@ -6102,9 +6112,12 @@ export function cleanUpMessage({ getMessage, isImpersonate, isContinue, displayI
         const nameToTrim2 = isImpersonate
-            ? (!power_user.allow_name1_display ? name1 : '')
-            : (!power_user.allow_name2_display ? name2 : '');
+            ? (!power_user.allow_name1_display ? name1 : '')  // user
```
The condition is correct, but it's reverted here and not in trimWrongNames case. Would be cooler to be consistent.

qvink replied:
done

qvink committed:
adding missing colon

qvink committed:
the cooler ternary

Cohee1207 commented on public/script.js:
```diff
@@ -5961,21 +6024,35 @@ export function cleanUpMessage(getMessage, isImpersonate, isContinue, displayInc
+            }
+
+            // remove any whitespace at the beginning
+            getMessage = getMessage.trimStart();
```
Unconditional space trimming is not needed here. There's a `power_user.trim_spaces;` condition at the end. Or put this under the `startIndex >= 0` condition.

qvink replied:
sure - I've removed it

qvink committed:
removing whitespace trimming in trimWrongNames block

Cohee1207 reviewed (approved):
We did it!
